cooperative_lanechange/src/cooperative_lanechange.cpp

```diff
@@ -0,0 +1,897 @@
+        
+        //Check if Lane Change is needed
+        std::vector<cav_msgs::Maneuver> maneuver_plan;
+        for(const auto maneuver : req.maneuver_plan.maneuvers)
+        {
+            if(maneuver.type == cav_msgs::Maneuver::LANE_CHANGE)
+            {
+                maneuver_plan.push_back(maneuver);
+            }
+        }
+
+        //Current only checking for first lane change maneuver message
+        int target_lanelet_id = stoi(maneuver_plan[0].lane_change_maneuver.ending_lane_id);
+        double target_downtrack = maneuver_plan[0].lane_change_maneuver.end_dist;
+        //get subject vehicle info
+        lanelet::BasicPoint2d veh_pos(req.vehicle_state.X_pos_global, req.vehicle_state.Y_pos_global);
+        double current_downtrack = wm_->routeTrackPos(veh_pos).downtrack;
+        auto current_lanelets = lanelet::geometry::findNearest(wm_->getMap()->laneletLayer, veh_pos, 10);       
+        int current_lanelet_id = current_lanelets[0].second.id();
+        if(current_lanelet_id == target_lanelet_id && current_downtrack >= target_downtrack - destination_range_){
+            cav_msgs::LaneChangeStatus lc_status_msg;
+            lc_status_msg.status = cav_msgs::LaneChangeStatus::PLANNING_SUCCESS;
+            //No description as per UI documentation
+            lanechange_status_pub_.publish(lc_status_msg);
+        }
+
+        int veh2_lanelet_id = 0;
+        double veh2_downtrack = 0.0, veh2_speed = 0.0;
+        bool foundRoadwayObject = false;
+        bool negotiate = true;
+        std::vector<cav_msgs::RoadwayObstacle> rwol = wm_->getRoadwayObjects();
+        //Assuming only one connected vehicle in list 
+        for(int i=0;i<rwol.size();i++){
+            if(rwol[i].connected_vehicle_type.type == cav_msgs::ConnectedVehicleType::CONNECTED_AND_AUTOMATED){
+                veh2_lanelet_id = rwol[0].lanelet_id;
+                veh2_downtrack = rwol[0].down_track; //Returns downtrack
+                veh2_speed = rwol[0].object.velocity.twist.linear.x;
+                foundRoadwayObject = true;
+                break;
+            }
+        }
+        if(foundRoadwayObject){
+            //get current_gap
+            double current_gap = find_current_gap(veh2_lanelet_id,veh2_downtrack);
+            //get desired gap - desired time gap (default 3s)* relative velocity
+            double relative_velocity = current_speed_ - veh2_speed;
+            double desired_gap = desired_time_gap_ * relative_velocity;      
+            
+            if(current_gap > desired_gap){
+                negotiate = false;  //No need for negotiation
+            }
+            
+        }
+        else{
+            ROS_WARN_STREAM("Did not find a connected and automated vehicle roadway object");
+        }
+
+        //plan lanechange without filling in response
+        std::vector<cav_msgs::TrajectoryPlanPoint> planned_trajectory_points = plan_lanechange(req);
+        
+        ros::Time request_sent_time;
```

Comment:
Might need to initialize this somehow, not sure what the default value for ros::Time is, but I can't imagine it's relevant if we're going to use this data element later.